velox/functions/lib/aggregates/MinMaxAggregateBase.cpp

```diff
@@ -0,0 +1,708 @@
+      mayPushdown = false;
+    }
+    if (mayPushdown && args[0]->isLazy()) {
+      BaseAggregate::template pushdown<MinMaxHook<T, true>>(
+          groups, rows, args[0]);
+      return;
+    }
+    BaseAggregate::template updateGroups<true, T>(
+        groups, rows, args[0], updateGroup, mayPushdown);
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    addRawInput(groups, rows, args, mayPushdown);
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    BaseAggregate::updateOneGroup(
+        group,
+        rows,
+        args[0],
+        updateGroup,
+        [](T& result, T value, int /* unused */) { result = value; },
+        mayPushdown,
+        kInitialValue_);
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    addSingleGroupRawInput(group, rows, args, mayPushdown);
+  }
+
+ protected:
+  static inline void updateGroup(T& result, T value) {
+    if constexpr (std::is_floating_point_v<T>) {
+      if (util::floating_point::NaNAwareGreaterThan<T>{}(result, value)) {
+        result = value;
+      }
+    } else {
+      if (result > value) {
+        result = value;
+      }
+    }
+  }
+
+  void initializeNewGroupsInternal(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    exec::Aggregate::setAllNulls(groups, indices);
+    for (auto i : indices) {
+      *exec::Aggregate::value<T>(groups[i]) = kInitialValue_;
+    }
+  }
+
+ private:
+  static const T kInitialValue_;
+};
+
+template <typename T>
+const T MinAggregate<T>::kInitialValue_ = MinMaxTrait<T>::max();
+
+// In velox, NaN is considered larger than infinity for floating point types.
+template <>
+const float MinAggregate<float>::kInitialValue_ =
+    MinMaxTrait<float>::quiet_NaN();
+
+template <>
+const double MinAggregate<double>::kInitialValue_ =
+    MinMaxTrait<double>::quiet_NaN();
+
+class TimestampMicrosPrecisionMaxAggregate : public MaxAggregate<Timestamp> {
+ public:
+  explicit TimestampMicrosPrecisionMaxAggregate(TypePtr resultType)
+      : MaxAggregate<Timestamp>(resultType) {}
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    using BaseAggregate =
+        SimpleNumericAggregate<Timestamp, Timestamp, Timestamp>;
+    BaseAggregate::template doExtractValues<Timestamp>(
+        groups, numGroups, result, [&](char* group) {
+          auto ts = *BaseAggregate::Aggregate::template value<Timestamp>(group);
+          return Timestamp::fromMicros(ts.toMicros());
+        });
+  }
+};
+
+class TimestampMicrosPrecisionMinAggregate : public MinAggregate<Timestamp> {
```

Comment:
This class is used by Spark only and does it make sense to put it in spark folder instead?

Reply:
`MinAggregate` is declared in anonymous namespace, this is the reason we alos declare `TimestampMicrosPrecisionMinAggregate` here. Added comments for these two classes.

Reply:
Since it is the lib, does it make sense to move MinAggregate out from anonymous namespace and use it in Spark folder?

Reply:
Moved `TimestampMicrosPrecisionMinAggregate` to spark folder, thanks.

Reply:
Thanks for the update. As quoted from Masha, APIs in 'detail' namespaces are not supposed to be used outside of the header/cpp files. Can we move it in the `functions::aggregate` namespace?

Reply:
Get it, updated, thanks.